Mathlib/Data/Fin/Tuple/Basic.lean

```diff
@@ -0,0 +1,898 @@
+/-
+Copyright (c) 2019 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn, Yury Kudryashov, Sébastien Gouëzel, Chris Hughes
+
+! This file was ported from Lean 3 source module data.fin.tuple.basic
+! leanprover-community/mathlib commit 7c523cb78f4153682c2929e3006c863bfef463d0
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Fin.Basic
+import Mathlib.Data.Pi.Lex
+import Mathlib.Data.Set.Intervals.Basic
+import Mathlib.Tactic.LibrarySearch -- **TODO** delete this later
+
+/-!
+# Operation on tuples
+
+We interpret maps `∀ i : Fin n, α i` as `n`-tuples of elements of possibly varying type `α i`,
+`(α 0, …, α (n-1))`. A particular case is `Fin n → α` of elements with all the same type.
+In this case when `α i` is a constant map, then tuples are isomorphic (but not definitionally equal)
+to `Vector`s.
+
+We define the following operations:
+
+* `Fin.tail` : the tail of an `n+1` tuple, i.e., its last `n` entries;
+* `Fin.cons` : adding an element at the beginning of an `n`-tuple, to get an `n+1`-tuple;
+* `Fin.init` : the beginning of an `n+1` tuple, i.e., its first `n` entries;
+* `Fin.snoc` : adding an element at the end of an `n`-tuple, to get an `n+1`-tuple. The name `snoc`
+  comes from `cons` (i.e., adding an element to the left of a tuple) read in reverse order.
+* `Fin.insertNth` : insert an element to a tuple at a given position.
+* `Fin.find p` : returns the first index `n` where `p n` is satisfied, and `none` if it is never
+  satisfied.
+
+-/
+
+set_option autoImplicit false -- **TODO** delete this later
+
+universe u v
+
+namespace Fin
+
+variable {m n : ℕ}
+
+open Function
+
+section Tuple
+
+/-- There is exactly one tuple of size zero. -/
+example (α : Fin 0 → Sort u) : Unique (∀ i : Fin 0, α i) := by infer_instance
+
+@[simp]
+theorem tuple0_le {α : ∀ _ : Fin 0, Type _} [∀ i, Preorder (α i)] (f g : ∀ i, α i) : f ≤ g :=
+  finZeroElim
+#align fin.tuple0_le Fin.tuple0_le
+
+variable {α : Fin (n + 1) → Type u} (x : α 0) (q : ∀ i, α i) (p : ∀ i : Fin n, α i.succ) (i : Fin n)
+  (y : α i.succ) (z : α 0)
+
+/-- The tail of an `n+1` tuple, i.e., its last `n` entries. -/
+def tail (q : ∀ i, α i) : ∀ i : Fin n, α i.succ := fun i ↦ q i.succ
+#align fin.tail Fin.tail
+
+theorem tail_def {n : ℕ} {α : Fin (n + 1) → Type _} {q : ∀ i, α i} :
+    (tail fun k : Fin (n + 1) ↦ q k) = fun k : Fin n ↦ q k.succ :=
+  rfl
+#align fin.tail_def Fin.tail_def
+
+-- Porting note: I made this noncomputable because Lean seemed to think it should be
+/-- Adding an element at the beginning of an `n`-tuple, to get an `n+1`-tuple. -/
+noncomputable def cons (x : α 0) (p : ∀ i : Fin n, α i.succ) : ∀ i, α i := fun j ↦ Fin.cases x p j
+#align fin.cons Fin.cons
```

Comment:
Is there a lean issue tracking this? It seems bizarre that `SomeInductive.cases` is noncomputable when the equation compiler is happy to produce something computable.

Reply:
The compiler ignores the actual Lean term generated by pattern matching and just compiles the pattern matching directly. I'm sure it's on the todo list of the developers.